### PR TITLE
gh-101021: Document binary parameters as bytes

### LINKED ISSUE #101021: The docs for email.mime types incorrectly identify bytes as strings
# Documentation

For example:
> _imagedata is a string containing the raw image data. ([image.py](https://github.com/python/cpython/blob/main/Lib/email/mime/image.py#L20))

or

> _audiodata is a string containing the raw audio data. ([audio.py](https://github.com/python/cpython/blob/main/Lib/email/mime/audio.py#L21))

So the programmer thinks, "how can it be a 'string' _and_ 'raw' at the same time?" Then the programmer notices the parameter for encoding and the language saying "the default encoding is Base64" and thinks, "ah, ok, that's how it could be a string!" (though it's still a little bit squirrelly to use the word "raw" to describe encoded data).

But then the reader sees the future tense in "... which will perform the actual encoding..." and thinks, "no, the argument must really be for binary data."

And then the programmer notices the parallel language in [text.py](https://github.com/python/cpython/blob/main/Lib/email/mime/text.py#L20):

> _text is the string for this message object.

and starts to wonder: "huh, maybe _all_ of these are expecting bytes." But then the programmer's code blows up when it tries to go with that theory (still no help from the documentation).

So in desperation, the programmer starts to dig into the actual code to find out what's done for each of these subclasses. And that journey leads to the conclusion that "string" in this package means one thing for some of the classes and something different for others.

Can't we just say "bytes" when we mean bytes, and reserve the use of the word "string" to mean what it's supposed to mean in the only versions of Python which are actually officially supported? 😆

### Linked PRs
* gh-101024
* gh-101043
* gh-101052

### PR DESCRIPTION
**Pull request by bkline**
The existing documentation refers to binary parameters as "strings" when they are actually bytes.

Fixes #101021

* Issue: gh-101021

### CODE REVIEW

File: Lib/email/mime/application.py
Comment (warsaw) [inline Lib/email/mime/application.py:20]: Operationally, it doesn't matter for `MIMEApplication`:

```
>>> from email.mime.application import MIMEApplication
>>> MIMEApplication('data')
<email.mime.application.MIMEApplication object at 0x1035e0e90>
>>> a = MIMEApplication('data')
>>> b = MIMEApplication(b'data')
>>> a.get_payload()
'ZGF0YQ==\n'
>>> b.get_payload()
'ZGF0YQ==\n'
```

although it still probably makes sense to promote bytes over strings here.

**Review by warsaw — CHANGES_REQUESTED** [review]
Can you resolve the conflicts and make sure all the checks pass?  Since this is a just a docstring change, adding the `skip-news` label is probably fine.

This could probably be backported to maintenance branches.

Have you checked the actual stdlib documentation for its language?

File: Lib/email/mime/application.py
Comment (bkline) [reply Lib/email/mime/application.py:20]: I'd be happy to make the requested changes if i knew how. If there's an obvious way to edit the labels on the PR I haven't found it. (I have a vague recollection of asking about this same thing last time I submitted a PR for a Python repo, and never getting an answer). As for whether I've "checked the actual stdlib documentation," I'm not really sure what that means. I mean, I assume it's obvious that I filed the ticket in the first place because I was reading the [documentation](https://docs.python.org/3/library/email.mime.html) for the Python standard library and found it confusing and misleading (and I provided links to the sections to which I was referring). So presumably that's not what you're asking.

Maybe I should just confine my contributions to filing the occasional bug report and leave the PRs to the insiders who actually know what they're doing. 🤔

Comment (warsaw) [reply Lib/email/mime/application.py:20]: > Maybe I should just confine my contributions to filing the occasional bug report and leave the PRs to the insiders who actually know what they're doing

You're doing great, and I thank you the effort to make this better for everyone!  I'm happy to help you get this one over the line.

I've gone ahead ahead and added the appropriate labels: `skip-news` as mentioned, but also backport labels to the [active maintenance (bug fix) releases](https://devguide.python.org/versions/), Python 3.10 and 3.11.

Comment (warsaw) [reply Lib/email/mime/application.py:20]: > As for whether I've "checked the actual stdlib documentation," I'm not really sure what that means.

Your PR touches the docstrings for three standard library files, which is great, but it doesn't actually update the documentation that gets published at the link you included.  Can you also update [this file](https://raw.githubusercontent.com/python/cpython/main/Doc/library/email.mime.rst)?

Comment (bkline) [reply Lib/email/mime/application.py:20]: Ah! So in spite of [this language](https://docs.python.org/3/glossary.html#term-docstring):

> **docstring**
> A string literal which appears as the first expression in a class, function or module. While ignored when the suite is executed, it is recognized by the compiler and put into the __doc__ attribute of the enclosing class, function or module. Since it is available via introspection, **_it is the canonical place for documentation of the object_**. [emphasis added]
> 

... the same language is being maintained in two different places? That's unfortunate. I'll add the fix to the second place.

As for the labels, I still haven't ever been given an answer to my original question: How does a contributor without commit access add a label to a PR? I've looked everywhere for the interface I would need and it's very elusive.

**Comment by bkline** [comment]
I have made the requested changes; please review again.

**Comment by bitdancer** [comment]
For your question about the labels, if you can't actually change labels using the 'labels' section in the right column of the PR on the github web interface, then you don't have permission to change those labels.  If I'm right about that then Barry probably just forgot that regular users can't make such changes.

On the subject of the documentation, the answer is that the Python documentation is not so much documentation of the cpython interpreter and the standard library as it is a *specification* of the Python language and the standard library.  (This fact is sometimes forgotten and results in later doc fixes prompted by people from other implementations of python noticing a statement that is cpython specific but not marked as such.)  While it would be technically possible to have the documentation for things written in python be docstrings extracted into the specification docs, it has so far been deemed undesirable to do that.  I think this is a combination of the aforementioned specification nature of the docs leading to a desire to be consistent (*all* the docs are in the doc source tree, not just some of them, to emphasize their independence from the specific implementation), and the fact that the language in the specification version of the docs, while often very similar or even identical to the function descriptions in the docstrings, is often enough different from what you would write if you were thinking of the function doc as a standalone docstring.  Including the fact that we prefer to not have sphinx markup in docstrings.

**Comment by bkline** [comment]
Thanks, David. Love your avatar.

Might be a good idea to teach the bots (if not the maintainers) not to nag contributors to do something they aren't allowed to do. 😉

I understand your explanation of how the documentation system works, and I agree that what you've described is probably the right approach. One of the costs of that approach, though, is that it increases the work for keeping the documentation current and in sync with itself (and the likelihood that problems like the one being fixed here will persist). Oh, well. Life involves tradeoffs. 😅

**Comment by warsaw** [comment]
I was trying to look through the dev guide to find a description of permissions for labeling.  I think the closest it gets is in the [triage team](https://devguide.python.org/triage/triage-team/#triage-team) description.  So yes, @bitdancer is probably right that @bkline doesn't have permission to add labels.

Normally in my own libraries, I extract API documentation from docstrings and type hints, but even with that, it's difficult to describe the APIs with the level of detail necessary for the standard library.

**Review by warsaw — APPROVED** [review]
Thanks again for your contribution @bkline.  Everything looks great.

**Comment by warsaw** [comment]
@bkline - as you can see, the automation couldn't backport the change.  If you're able to do that and submit a PR, I can help get that landed.  If you can't or don't want to, let me know and I'll try to get to it when I'm back on my laptop.

**Comment by bkline** [comment]
I'd be willing to give it a shot, given adequate instructions. Not 100% sure I'd know all the steps which are needed otherwise. Fortunately, Google has gotten a lot better about bringing up the most current version of the docs in response to searches, so it's not as critical as the backport to 3.10. Short answer: we're probably better off if you do it. 😎

**Comment by warsaw** [comment]
@bkline The devguide does explain [how to manually backport](https://devguide.python.org/getting-started/git-boot-camp/#backporting-merged-changes) changes.  It's entirely up to you - if you want to learn how to do it and it would be fun for you, go for it!  Just tag me in the PR.  If not, I'll give it a shot.

**Comment by warsaw** [comment]
Now that I'm on my desktop, I can see that the 3.11 backport worked fine.  I've approved and merged it.  It's only the 3.10 backport that failed.  I'm looking at that now.

**Comment by bkline** [comment]
@warsaw Yes, those look like excellent instructions. I will definitely give it a shot.

**Comment by bkline** [comment]
Well, not as much fun as we thought it would be.
```
(venv) √ mime % cherry_picker --dry-run 49cae39ef020eaf242607bb2d2d193760b9855a6 3.10
🐍 🍒 ⛏
Dry run requested, listing expected command sequence
  dry-run: git remote get-url upstream
  dry-run: git fetch upstream --no-tags
Now backporting '49cae39ef020eaf242607bb2d2d193760b9855a6' into '3.10'
  dry-run: git remote get-url upstream
  dry-run: git checkout -b backport-49cae39-3.10 upstream/3.10
  dry-run: git cherry-pick -x 49cae39ef020eaf242607bb2d2d193760b9855a6

  dry-run: git show -s --format=%B 49cae39ef020eaf242607bb2d2d193760b9855a6
Traceback (most recent call last):
  File "/Users/bkline/.venvs/cdr/bin/cherry_picker", line 8, in <module>
    sys.exit(cherry_pick_cli())
  File "/Users/bkline/.venvs/cdr/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/bkline/.venvs/cdr/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/bkline/.venvs/cdr/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/bkline/.venvs/cdr/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/bkline/.venvs/cdr/lib/python3.10/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/bkline/.venvs/cdr/lib/python3.10/site-packages/cherry_picker/cherry_picker.py", line 645, in cherry_pick_cli
    cherry_picker.backport()
  File "/Users/bkline/.venvs/cdr/lib/python3.10/site-packages/cherry_picker/cherry_picker.py", line 387, in backport
    commit_message = self.amend_commit_message(cherry_pick_branch)
  File "/Users/bkline/.venvs/cdr/lib/python3.10/site-packages/cherry_picker/cherry_picker.py", line 269, in amend_commit_message
    updated_commit_message = f"""{commit_prefix}{self.get_commit_message(self.commit_sha1)}
  File "/Users/bkline/.venvs/cdr/lib/python3.10/site-packages/cherry_picker/cherry_picker.py", line 211, in get_commit_message
    message = self.run_cmd(cmd).strip()
AttributeError: 'NoneType' object has no attribute 'strip'
```
I think I'll leave this to the big dogs. 😆🐾

**Comment by warsaw** [comment]
> I think I'll leave this to the big dogs

Yeah, I'm sorry about all the friction.  I ran into a few things myself so at least the core devs have also had a good discussion about this experience.

That all aside, I was able to manually backport the change the 3.10, so we're all done here.  Thanks again for your contribution!

**Comment by bkline** [comment]
No worries, and thanks for all your help (and for helping to keep Python the best of the many languages I've used over the decades).

**Pull request merged**
